{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21f3eaab-0d93-4cdb-b626-e63e02d1c41a",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-11-21T11:33:32.632256Z",
     "iopub.status.busy": "2023-11-21T11:33:32.631561Z",
     "iopub.status.idle": "2023-11-21T11:33:35.429907Z",
     "shell.execute_reply": "2023-11-21T11:33:35.428896Z",
     "shell.execute_reply.started": "2023-11-21T11:33:32.632213Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\woloo\\anaconda3\\envs\\sc\\Lib\\site-packages\\umap\\distances.py:1063: NumbaDeprecationWarning: \u001b[1mThe 'nopython' keyword argument was not supplied to the 'numba.jit' decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.\u001b[0m\n",
      "  @numba.jit()\n",
      "c:\\Users\\woloo\\anaconda3\\envs\\sc\\Lib\\site-packages\\umap\\distances.py:1071: NumbaDeprecationWarning: \u001b[1mThe 'nopython' keyword argument was not supplied to the 'numba.jit' decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.\u001b[0m\n",
      "  @numba.jit()\n",
      "c:\\Users\\woloo\\anaconda3\\envs\\sc\\Lib\\site-packages\\umap\\distances.py:1086: NumbaDeprecationWarning: \u001b[1mThe 'nopython' keyword argument was not supplied to the 'numba.jit' decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.\u001b[0m\n",
      "  @numba.jit()\n",
      "c:\\Users\\woloo\\anaconda3\\envs\\sc\\Lib\\site-packages\\tqdm\\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n",
      "c:\\Users\\woloo\\anaconda3\\envs\\sc\\Lib\\site-packages\\umap\\umap_.py:660: NumbaDeprecationWarning: \u001b[1mThe 'nopython' keyword argument was not supplied to the 'numba.jit' decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.\u001b[0m\n",
      "  @numba.jit()\n",
      "cffi mode is CFFI_MODE.ANY\n",
      "R home found: C:\\Program Files\\R\\R-4.3.3\n",
      "Default options to initialize R: rpy2, --quiet, --no-save\n",
      "R[write to console]: 载入需要的程辑包：SingleCellExperiment\n",
      "\n",
      "R[write to console]: 载入需要的程辑包：SummarizedExperiment\n",
      "\n",
      "R[write to console]: 载入需要的程辑包：MatrixGenerics\n",
      "\n",
      "R[write to console]: 载入需要的程辑包：matrixStats\n",
      "\n",
      "R[write to console]: \n",
      "载入程辑包：'MatrixGenerics'\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from 'package:matrixStats':\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "R[write to console]: 载入需要的程辑包：GenomicRanges\n",
      "\n",
      "R[write to console]: 载入需要的程辑包：stats4\n",
      "\n",
      "R[write to console]: 载入需要的程辑包：BiocGenerics\n",
      "\n",
      "R[write to console]: \n",
      "载入程辑包：'BiocGenerics'\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from 'package:stats':\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from 'package:base':\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "R[write to console]: 载入需要的程辑包：S4Vectors\n",
      "\n",
      "R[write to console]: \n",
      "载入程辑包：'S4Vectors'\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from 'package:utils':\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from 'package:base':\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "R[write to console]: 载入需要的程辑包：IRanges\n",
      "\n",
      "R[write to console]: \n",
      "载入程辑包：'IRanges'\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from 'package:grDevices':\n",
      "\n",
      "    windows\n",
      "\n",
      "\n",
      "R[write to console]: 载入需要的程辑包：GenomeInfoDb\n",
      "\n",
      "R[write to console]: 载入需要的程辑包：Biobase\n",
      "\n",
      "R[write to console]: Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "R[write to console]: \n",
      "载入程辑包：'Biobase'\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from 'package:MatrixGenerics':\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from 'package:matrixStats':\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "R[write to console]: In addition: \n",
      "R[write to console]: Warning messages:\n",
      "\n",
      "R[write to console]: 1: 程辑包'scDblFinder'是用R版本4.4.0 来建造的 \n",
      "\n",
      "R[write to console]: 2: 程辑包'SingleCellExperiment'是用R版本4.4.0 来建造的 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "import scanpy as sc\n",
    "import anndata as ad\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from scipy.stats import median_abs_deviation\n",
    "import seaborn as sns\n",
    "import celltypist\n",
    "from celltypist import models\n",
    "import scrublet as scr\n",
    "from scipy.sparse import csr_matrix\n",
    "import os\n",
    "\n",
    "def is_outlier(adata, metric: str, nmads: int):\n",
    "    M = adata.obs[metric]\n",
    "    outlier = (M < np.median(M) - nmads * median_abs_deviation(M)) | (\n",
    "        np.median(M) + nmads * median_abs_deviation(M) < M\n",
    "    )\n",
    "    return outlier\n",
    "\n",
    "# Sample No. of all human samples.\n",
    "human_datasets = [\"GSM4837523\",\"GSM4837524\",\"GSM4837525\",\"GSM4837526\",\"GSM4837527\",\"GSM4837528\",\n",
    "                  \n",
    "                  \"GSM8029950\",\"GSM8029951\",\"GSM8029952\",\"GSM8029953\",\"GSM8029954\",\"GSM8029955\",\n",
    "                  \"GSM8029956\",\"GSM8029957\",\"GSM8029958\",\"GSM8029959\",\"GSM8029960\",\"GSM8029961\",\n",
    "                  \"GSM8029962\",\"GSM8029963\",\"GSM8029964\",\"GSM8029965\",\"GSM8029966\",\"GSM8029967\",\n",
    "                  \"GSM8029968\",\"GSM8029969\",\"GSM8029970\",\"GSM8029971\",\"GSM8029972\",\"GSM8029973\",\n",
    "                  \n",
    "                  \"GSM8121894\",\"GSM8121895\",\"GSM8121896\",\"GSM8121897\",\"GSM8121898\",\"GSM8121899\",\n",
    "                  \"GSM8121900\",\"GSM8121901\",\"GSM8121902\",\"GSM8121903\",\"GSM8121904\",\"GSM8121905\",\n",
    "                  \"GSM8121906\",\"GSM8121907\",\"GSM8121908\",\"GSM5392187\",\"GSM5392188\",\"GSM5392189\",\n",
    "                  \n",
    "                  \"GSM5410851\",\"GSM5410853\",\"GSM5410854\",\"GSM5410855\",\"GSM5410856\",\"GSM6976304\",\n",
    "                  \"GSM6976305\",\"GSM6976306\",\"GSM6976307\",\"GSM6976308\",\"GSM6976309\",\"GSM6976310\",\n",
    "                  \"GSM6976311\",\"GSM6976312\",\"GSM6976313\",\n",
    "                  \n",
    "                  \"GSM6422822\",\"GSM6422823\",\"GSM6422824\",\"GSM6422825\",\"GSM6422826\",\"GSM6422827\",\n",
    "                  \n",
    "                  \"GSM3449613\",\"GSM3449614\",\"GSM3449619\",\"GSM3449620\",\n",
    "                  \n",
    "                  \"GSM4307515\",\"GSM4307516\",\"GSM4307517\",\"GSM4307518\",\"GSM4307519\",\"GSM4307530\",\n",
    "                  \"GSM4307531\",\"GSM4307532\",\"GSM4307533\",\"GSM4307534\",\"GSM4307535\",\"GSM4307536\",\n",
    "                  \"GSM4307537\",\"GSM4307538\",\"GSM4307539\",\"GSM4307540\",\"GSM4307541\",\"GSM4307542\",\n",
    "                  \"GSM4307543\",\"GSM4307544\",\"GSM4307545\",\"GSM4307551\",\"GSM4307552\",\n",
    "                  \n",
    "                  'GSM4705589','GSM4705590','GSM4705591',\n",
    "                  \n",
    "                  \"GSM5577199\",\"GSM5577200\",\"GSE131778\",\"GSE145154\",\"GSE155512\",\"GSE196943\",\n",
    "                  \n",
    "                  \"GSM3819856\",\"GSM3819857\",\"GSM3819858\",\"GSM3819859\",\"GSM3819860\",\"GSM3819861\",\n",
    "                  \"GSM3819862\",\"GSM3819863\",\n",
    "\n",
    "                  \"GSM5905363\",\"GSM5905364\",\"GSM5905365\",\"GSM5905366\",\"GSM5905367\",\"GSM5905368\",\n",
    "                  \"GSM5905370\",\"GSM5905371\",\"GSM5905372\",\"GSM5905373\",\"GSM5905375\",\"GSM5905377\"\n",
    "]\n",
    "\n",
    "# Load py to R interface\n",
    "import rpy2.rinterface_lib.callbacks as rcb\n",
    "import rpy2.robjects as ro\n",
    "import anndata2ri\n",
    "ro.pandas2ri.activate()\n",
    "anndata2ri.activate()\n",
    "ro.r('library(scDblFinder)')\n",
    "\n",
    "# path to the location where all folders named \"GSEXXXXXX\", \"E-MTAB-XXXX\", etc., exist. **Please end with \"/\"\n",
    "main_path = \"path/to/all/datasets/\"\n",
    "\n",
    "hen2sy = pd.read_table(main_path + \"human_annotation.txt\")\n",
    "hen2sy = dict(list(zip(hen2sy[\"Gene stable ID\"], hen2sy[\"Gene name\"])))\n",
    "men2sy = pd.read_table(main_path + \"mouse_annotation.txt\")\n",
    "men2sy = dict(list(zip(men2sy[\"Gene stable ID\"], men2sy[\"Gene name\"])))\n",
    "\n",
    "# lists of datasets with different file formats\n",
    "raw_10x_mtx = [\"GSE214611\",\"GSE159677\",\"GSE205930\",\"GSE150140\",\"GSE169332\",\"GSE197853\"] #6\n",
    "raw_csv = [\"GSE185265\",\"GSE227088\"] #2\n",
    "raw_txt = [\"GSE260656\",\"GSE260657\",\"GSE235275\",\"GSE179159\",\"GSE155512\",\"GSE155513\"] #5\n",
    "raw_tiled_10x_mtx = [\"GSE178469\",\"GSE201947\",\"GSE216211\",\"GSE253902\",\"GSE253903\",\"GSE246779\",\"GSE210159\",\"GSE106472\",\"E-MTAB-9816\",\"E-MTAB-9817\",\n",
    "                     \"GSE119355\",\"GSE128509\",\"GSE135310\",\"GSE145154\",\"GSE153480\",\"GSE163465\",\"GSE157244\",\"GSE184073\",\"GSE130699\",\"E-MTAB-9817\",\n",
    "                     \"GSE193426\"] #17\n",
    "raw_h5 = [\"GSE206787\",\"GSE236609\",\"GSE132144\",\"GSE163129\"] #4\n",
    "raw_h5ad = [\"GSE146285\",\"GSE121893\",\"GSE210152\",\"E-MTAB-7376\",\"GSE196943\",\"GSE207275\",\"GSE135296\",\"GSE131776\",\"GSE131778\",\"E-MTAB-9583\",\"GSE197441\"\n",
    "            ] #10\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa1053b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# For all samples\n",
    "''''''\n",
    "for a in raw_10x_mtx:\n",
    "    # for expression matrices of 10x raw format\n",
    "    dataset_id = a\n",
    "    path_li = list(os.walk(main_path+dataset_id+\"_RAW/\"))[1:]\n",
    "    print(\"\\033[33mReading dataset {} ...\\033[0m\".format(a))\n",
    "    for j in path_li:\n",
    "        \n",
    "        adata = sc.read_10x_mtx(j[0], prefix=\"\", gex_only=\"False\")\n",
    "        j = j[0].split(\"/\")[-1]\n",
    "        print(\"\\033[32mReading sample {} ...\\033[0m\".format(j))\n",
    "        \n",
    "        # convert ensembl ids to gene symbols\n",
    "        new_var = list(adata.var_names)\n",
    "        if j in human_datasets:\n",
    "            new_var = [(\"MT-\"+i.split(\"MT\")[1]) if (i.startswith(\"MT\") and not i.startswith(\"MT-\")) else i for i in new_var] # turn \"MTXXX\" into \"MT-XXX\"\n",
    "            for i in range(len(new_var)):\n",
    "                try:\n",
    "                    if new_var[i][:4]==\"ENSG\" and pd.isna(hen2sy[new_var[i]])==False:\n",
    "                        new_var[i] = hen2sy[new_var[i]]\n",
    "                except:\n",
    "                    continue\n",
    "        else:\n",
    "            new_var = [(\"mt-\"+i.split(\"mt\")[1]) if (i.startswith(\"mt\") and not i.startswith(\"mt-\")) else i for i in new_var] # turn \"mtXXX\" into \"mt-XXX\"\n",
    "            for i in range(len(new_var)):\n",
    "                # if mouse mito gene symbols were in upper case, turn into lower case with \"mt-\"\n",
    "                if new_var[i] in [n.upper() for n in [\"Tf\",\"Rnr1\",\"Tv\",\"Rnr2\",\"Tl1\",\"Nd1\",\"Ti\",\"Tq\",\"Tm\",\"Nd2\",\"Tw\",\"Ta\",\"Tn\",\n",
    "                                                        \"Tc\",\"Ty\",\"Co1\",\"Ts1\",\"Td\",\"Co2\",\"Tk\",\"Atp8\",\"Atp6\",\"Co3\",\"Tg\",\"Nd3\",\"Tr\",\n",
    "                                                        \"Nd4l\",\"Nd4\",\"Th\",\"Ts2\",\"Tl2\",\"Nd5\",\"Nd6\",\"Te\",\"Cytb\",\"Tt\",\"Tp\"]]:\n",
    "                    new_var[i] = \"mt-\" + new_var[i][0] + new_var[i][1:].lower()\n",
    "                # turn ensembl id into gene symbol\n",
    "                try:\n",
    "                    if new_var[i][:7]==\"ENSMUSG\" and pd.isna(men2sy[new_var[i]])==False:\n",
    "                        new_var[i] = men2sy[new_var[i]]\n",
    "                except:\n",
    "                    continue\n",
    "        adata.var_names = pd.Index(new_var)\n",
    "        \n",
    "        # remove low sequencing quality cells\n",
    "        adata.var_names_make_unique()\n",
    "        adata.obs_names_make_unique()\n",
    "        sc.pp.filter_cells(adata, min_genes=200)\n",
    "        if j in human_datasets:\n",
    "            adata.var[\"mt\"] = adata.var_names.str.startswith(\"MT-\")\n",
    "        else:\n",
    "            adata.var[\"mt\"] = adata.var_names.str.startswith(\"mt-\") | adata.var_names.isin([n.upper() for n in [\"Tf\",\"Rnr1\",\"Tv\",\"Rnr2\",\"Tl1\",\"Nd1\",\"Ti\",\"Tq\",\"Tm\",\"Nd2\",\"Tw\",\"Ta\",\"Tn\",\n",
    "                                                                                                                \"Tc\",\"Ty\",\"Co1\",\"Ts1\",\"Td\",\"Co2\",\"Tk\",\"Atp8\",\"Atp6\",\"Co3\",\"Tg\",\"Nd3\",\"Tr\",\n",
    "                                                                                                                \"Nd4l\",\"Nd4\",\"Th\",\"Ts2\",\"Tl2\",\"Nd5\",\"Nd6\",\"Te\",\"Cytb\",\"Tt\",\"Tp\"]])\n",
    "        sc.pp.calculate_qc_metrics(\n",
    "            adata, qc_vars=[\"mt\"], inplace=True, percent_top=[20], log1p=True\n",
    "        )\n",
    "        adata.obs[\"outlier\"] = (\n",
    "            is_outlier(adata, \"log1p_total_counts\", 5)\n",
    "            | is_outlier(adata, \"log1p_n_genes_by_counts\", 5)\n",
    "            | is_outlier(adata, \"pct_counts_in_top_20_genes\", 5)\n",
    "        )\n",
    "        adata.obs.outlier.value_counts()\n",
    "        adata.obs[\"mt_outlier\"] = is_outlier(adata, \"pct_counts_mt\", 5)\n",
    "        adata.obs.mt_outlier.value_counts()\n",
    "        adata.obs[\"is_outlier\"] = [\"outlier\" if i else \"non-outlier\" for i in adata.obs[\"outlier\"]]\n",
    "        adata.obs[\"is_mt_outlier\"] = [\"mt_outlier\" if i else \"non-mt_outlier\" for i in adata.obs[\"mt_outlier\"]]\n",
    "        sc.pl.violin(adata, [\"pct_counts_mt\"], groupby = \"is_mt_outlier\")\n",
    "        \n",
    "        print(f\"Total number of cells: {adata.n_obs}\")\n",
    "        adata = adata[(~adata.obs.outlier)\\\n",
    "                    & (~adata.obs.mt_outlier)\\\n",
    "                    ].copy()\n",
    "        \n",
    "        print(f\"Number of cells after filtering of low quality cells: {adata.n_obs}\")\n",
    "        counts_matrix = adata.X\n",
    "        ro.globalenv[\"adata\"] = counts_matrix.T.todense()\n",
    "        ro.r('doublets <- scDblFinder(sce = as.matrix(adata), verbose=FALSE)')\n",
    "        doublet_scores = ro.r('doublets$scDblFinder.score').T\n",
    "        doublets = ro.r('doublets$scDblFinder.class')\n",
    "        adata.obs[\"doublet_score\"] = doublet_scores\n",
    "        adata.obs[\"doublet\"] = doublets\n",
    "        adata.obs[\"doublet\"] = [\"doublet\" if i!=1 else \"singlet\" for i in adata.obs[\"doublet\"]]\n",
    "        adata.obs[\"is_doublet\"] = [True if i==\"doublet\" else False for i in adata.obs[\"doublet\"]]\n",
    "        \n",
    "        print(f\"Total number of cells: {adata.n_obs}\")\n",
    "        adata = adata[(~adata.obs.is_doublet)].copy()\n",
    "\n",
    "        print(f\"Number of cells after filtering of doublets: {adata.n_obs}\")\n",
    "        adata.write(main_path+\"datanew/\"+j+\".h5ad\")\n",
    "\n",
    "for a in raw_csv:\n",
    "    # For CSV format\n",
    "    dataset_id = a\n",
    "    path_li = list(os.walk(main_path+dataset_id+\"_RAW/\"))[0][2]\n",
    "    print(\"\\033[33mReading dataset {} ...\\033[0m\".format(a))\n",
    "    for j in path_li:\n",
    "        \n",
    "        print(\"\\033[33mReading sample {} ...\\033[0m\".format(j))\n",
    "        adata = sc.read_text(main_path+dataset_id+\"_RAW/\"+j, delimiter=\",\").T\n",
    "        j = j.split(\"_\")[0].split(\".\")[0]\n",
    "        \n",
    "        # convert ensembl ids to gene symbols\n",
    "        new_var = list(adata.var_names)\n",
    "        if j in human_datasets:\n",
    "            new_var = [(\"MT-\"+i.split(\"MT\")[1]) if (i.startswith(\"MT\") and not i.startswith(\"MT-\")) else i for i in new_var] # turn \"MTXXX\" into \"MT-XXX\"\n",
    "            for i in range(len(new_var)):\n",
    "                try:\n",
    "                    if new_var[i][:4]==\"ENSG\" and pd.isna(hen2sy[new_var[i]])==False:\n",
    "                        new_var[i] = hen2sy[new_var[i]]\n",
    "                except:\n",
    "                    continue\n",
    "        else:\n",
    "            new_var = [(\"mt-\"+i.split(\"mt\")[1]) if (i.startswith(\"mt\") and not i.startswith(\"mt-\")) else i for i in new_var] # turn \"mtXXX\" into \"mt-XXX\"\n",
    "            for i in range(len(new_var)):\n",
    "                # if mouse mito gene symbols were in upper case, turn into lower case with \"mt-\"\n",
    "                if new_var[i] in [n.upper() for n in [\"Tf\",\"Rnr1\",\"Tv\",\"Rnr2\",\"Tl1\",\"Nd1\",\"Ti\",\"Tq\",\"Tm\",\"Nd2\",\"Tw\",\"Ta\",\"Tn\",\n",
    "                                                        \"Tc\",\"Ty\",\"Co1\",\"Ts1\",\"Td\",\"Co2\",\"Tk\",\"Atp8\",\"Atp6\",\"Co3\",\"Tg\",\"Nd3\",\"Tr\",\n",
    "                                                        \"Nd4l\",\"Nd4\",\"Th\",\"Ts2\",\"Tl2\",\"Nd5\",\"Nd6\",\"Te\",\"Cytb\",\"Tt\",\"Tp\"]]:\n",
    "                    new_var[i] = \"mt-\" + new_var[i][0] + new_var[i][1:].lower()\n",
    "                # turn ensembl id into gene symbol\n",
    "                try:\n",
    "                    if new_var[i][:7]==\"ENSMUSG\" and pd.isna(men2sy[new_var[i]])==False:\n",
    "                        new_var[i] = men2sy[new_var[i]]\n",
    "                except:\n",
    "                    continue\n",
    "        adata.var_names = pd.Index(new_var)\n",
    "        \n",
    "        # remove low sequencing quality cells\n",
    "        adata.var_names_make_unique()\n",
    "        adata.obs_names_make_unique()\n",
    "        sc.pp.filter_cells(adata, min_genes=200)\n",
    "        if j in human_datasets:\n",
    "            adata.var[\"mt\"] = adata.var_names.str.startswith(\"MT-\")\n",
    "        else:\n",
    "            adata.var[\"mt\"] = adata.var_names.str.startswith(\"mt-\") | adata.var_names.isin([n.upper() for n in [\"Tf\",\"Rnr1\",\"Tv\",\"Rnr2\",\"Tl1\",\"Nd1\",\"Ti\",\"Tq\",\"Tm\",\"Nd2\",\"Tw\",\"Ta\",\"Tn\",\n",
    "                                                                                                                \"Tc\",\"Ty\",\"Co1\",\"Ts1\",\"Td\",\"Co2\",\"Tk\",\"Atp8\",\"Atp6\",\"Co3\",\"Tg\",\"Nd3\",\"Tr\",\n",
    "                                                                                                                \"Nd4l\",\"Nd4\",\"Th\",\"Ts2\",\"Tl2\",\"Nd5\",\"Nd6\",\"Te\",\"Cytb\",\"Tt\",\"Tp\"]])\n",
    "        sc.pp.calculate_qc_metrics(\n",
    "            adata, qc_vars=[\"mt\"], inplace=True, percent_top=[20], log1p=True\n",
    "        )\n",
    "        adata.obs[\"outlier\"] = (\n",
    "            is_outlier(adata, \"log1p_total_counts\", 5)\n",
    "            | is_outlier(adata, \"log1p_n_genes_by_counts\", 5)\n",
    "            | is_outlier(adata, \"pct_counts_in_top_20_genes\", 5)\n",
    "        )\n",
    "        adata.obs.outlier.value_counts()\n",
    "        adata.obs[\"mt_outlier\"] = is_outlier(adata, \"pct_counts_mt\", 5)\n",
    "        adata.obs.mt_outlier.value_counts()\n",
    "        adata.obs[\"is_outlier\"] = [\"outlier\" if i else \"non-outlier\" for i in adata.obs[\"outlier\"]]\n",
    "        adata.obs[\"is_mt_outlier\"] = [\"mt_outlier\" if i else \"non-mt_outlier\" for i in adata.obs[\"mt_outlier\"]]\n",
    "        sc.pl.violin(adata, [\"pct_counts_mt\"], groupby = \"is_mt_outlier\")\n",
    "        \n",
    "        print(f\"Total number of cells: {adata.n_obs}\")\n",
    "        adata = adata[(~adata.obs.outlier)\\\n",
    "                    & (~adata.obs.mt_outlier)\\\n",
    "                    ].copy()\n",
    "        \n",
    "        print(f\"Number of cells after filtering of low quality cells: {adata.n_obs}\")\n",
    "        counts_matrix = adata.X\n",
    "        ro.globalenv[\"adata\"] = counts_matrix.T\n",
    "        ro.r('doublets <- scDblFinder(sce = as.matrix(adata), verbose=FALSE)')\n",
    "        doublet_scores = ro.r('doublets$scDblFinder.score').T\n",
    "        doublets = ro.r('doublets$scDblFinder.class')\n",
    "        adata.obs[\"doublet_score\"] = doublet_scores\n",
    "        adata.obs[\"doublet\"] = doublets\n",
    "        adata.obs[\"doublet\"] = [\"doublet\" if i!=1 else \"singlet\" for i in adata.obs[\"doublet\"]]\n",
    "        adata.obs[\"is_doublet\"] = [True if i==\"doublet\" else False for i in adata.obs[\"doublet\"]]\n",
    "        \n",
    "        print(f\"Total number of cells: {adata.n_obs}\")\n",
    "        adata = adata[(~adata.obs.is_doublet)].copy()\n",
    "\n",
    "        print(f\"Number of cells after filtering of doublets: {adata.n_obs}\")\n",
    "        adata.write(main_path+\"datanew/\"+j+\".h5ad\")\n",
    "\n",
    "for a in raw_txt:\n",
    "    # For TXT format\n",
    "    dataset_id = a\n",
    "    path_li = list(os.walk(main_path+dataset_id+\"_RAW/\"))[0][2]\n",
    "    print(\"\\033[33mReading dataset {} ...\\033[0m\".format(a))\n",
    "    for j in path_li:\n",
    "        print(\"\\033[33mReading sample {} ...\\033[0m\".format(j))\n",
    "        adata = sc.read_text(main_path+dataset_id+\"_RAW/\"+j).T\n",
    "        j = j.split(\"_\")[0].split(\".\")[0]\n",
    "        \n",
    "        # convert ensembl ids to gene symbols\n",
    "        new_var = list(adata.var_names)\n",
    "        if j in human_datasets:\n",
    "            new_var = [(\"MT-\"+i.split(\"MT\")[1]) if (i.startswith(\"MT\") and not i.startswith(\"MT-\")) else i for i in new_var] # turn \"MTXXX\" into \"MT-XXX\"\n",
    "            for i in range(len(new_var)):\n",
    "                try:\n",
    "                    if new_var[i][:4]==\"ENSG\" and pd.isna(hen2sy[new_var[i]])==False:\n",
    "                        new_var[i] = hen2sy[new_var[i]]\n",
    "                except:\n",
    "                    continue\n",
    "        else:\n",
    "            new_var = [(\"mt-\"+i.split(\"mt\")[1]) if (i.startswith(\"mt\") and not i.startswith(\"mt-\")) else i for i in new_var] # turn \"mtXXX\" into \"mt-XXX\"\n",
    "            for i in range(len(new_var)):\n",
    "                # if mouse mito gene symbols were in upper case, turn into lower case with \"mt-\"\n",
    "                if new_var[i] in [n.upper() for n in [\"Tf\",\"Rnr1\",\"Tv\",\"Rnr2\",\"Tl1\",\"Nd1\",\"Ti\",\"Tq\",\"Tm\",\"Nd2\",\"Tw\",\"Ta\",\"Tn\",\n",
    "                                                        \"Tc\",\"Ty\",\"Co1\",\"Ts1\",\"Td\",\"Co2\",\"Tk\",\"Atp8\",\"Atp6\",\"Co3\",\"Tg\",\"Nd3\",\"Tr\",\n",
    "                                                        \"Nd4l\",\"Nd4\",\"Th\",\"Ts2\",\"Tl2\",\"Nd5\",\"Nd6\",\"Te\",\"Cytb\",\"Tt\",\"Tp\"]]:\n",
    "                    new_var[i] = \"mt-\" + new_var[i][0] + new_var[i][1:].lower()\n",
    "                # turn ensembl id into gene symbol\n",
    "                try:\n",
    "                    if new_var[i][:7]==\"ENSMUSG\" and pd.isna(men2sy[new_var[i]])==False:\n",
    "                        new_var[i] = men2sy[new_var[i]]\n",
    "                except:\n",
    "                    continue\n",
    "        adata.var_names = pd.Index(new_var)\n",
    "        \n",
    "        # remove low sequencing quality cells\n",
    "        adata.var_names_make_unique()\n",
    "        adata.obs_names_make_unique()\n",
    "        sc.pp.filter_cells(adata, min_genes=200)\n",
    "        if j in human_datasets:\n",
    "            adata.var[\"mt\"] = adata.var_names.str.startswith(\"MT-\")\n",
    "        else:\n",
    "            adata.var[\"mt\"] = adata.var_names.str.startswith(\"mt-\") | adata.var_names.isin([n.upper() for n in [\"Tf\",\"Rnr1\",\"Tv\",\"Rnr2\",\"Tl1\",\"Nd1\",\"Ti\",\"Tq\",\"Tm\",\"Nd2\",\"Tw\",\"Ta\",\"Tn\",\n",
    "                                                                                                                \"Tc\",\"Ty\",\"Co1\",\"Ts1\",\"Td\",\"Co2\",\"Tk\",\"Atp8\",\"Atp6\",\"Co3\",\"Tg\",\"Nd3\",\"Tr\",\n",
    "                                                                                                                \"Nd4l\",\"Nd4\",\"Th\",\"Ts2\",\"Tl2\",\"Nd5\",\"Nd6\",\"Te\",\"Cytb\",\"Tt\",\"Tp\"]])\n",
    "        sc.pp.calculate_qc_metrics(\n",
    "            adata, qc_vars=[\"mt\"], inplace=True, percent_top=[20], log1p=True\n",
    "        )\n",
    "        adata.obs[\"outlier\"] = (\n",
    "            is_outlier(adata, \"log1p_total_counts\", 5)\n",
    "            | is_outlier(adata, \"log1p_n_genes_by_counts\", 5)\n",
    "            | is_outlier(adata, \"pct_counts_in_top_20_genes\", 5)\n",
    "        )\n",
    "        adata.obs.outlier.value_counts()\n",
    "        adata.obs[\"mt_outlier\"] = is_outlier(adata, \"pct_counts_mt\", 5)\n",
    "        adata.obs.mt_outlier.value_counts()\n",
    "        adata.obs[\"is_outlier\"] = [\"outlier\" if i else \"non-outlier\" for i in adata.obs[\"outlier\"]]\n",
    "        adata.obs[\"is_mt_outlier\"] = [\"mt_outlier\" if i else \"non-mt_outlier\" for i in adata.obs[\"mt_outlier\"]]\n",
    "        sc.pl.violin(adata, [\"pct_counts_mt\"], groupby = \"is_mt_outlier\")\n",
    "        \n",
    "        print(f\"Total number of cells: {adata.n_obs}\")\n",
    "        adata = adata[(~adata.obs.outlier)\\\n",
    "                    & (~adata.obs.mt_outlier)\\\n",
    "                    ].copy()\n",
    "        \n",
    "        print(f\"Number of cells after filtering of low quality cells: {adata.n_obs}\")\n",
    "        counts_matrix = adata.X\n",
    "        ro.globalenv[\"adata\"] = counts_matrix.T\n",
    "        ro.r('doublets <- scDblFinder(sce = as.matrix(adata), verbose=FALSE)')\n",
    "        doublet_scores = ro.r('doublets$scDblFinder.score').T\n",
    "        doublets = ro.r('doublets$scDblFinder.class')\n",
    "        adata.obs[\"doublet_score\"] = doublet_scores\n",
    "        adata.obs[\"doublet\"] = doublets\n",
    "        adata.obs[\"doublet\"] = [\"doublet\" if i!=1 else \"singlet\" for i in adata.obs[\"doublet\"]]\n",
    "        adata.obs[\"is_doublet\"] = [True if i==\"doublet\" else False for i in adata.obs[\"doublet\"]]\n",
    "        \n",
    "        print(f\"Total number of cells: {adata.n_obs}\")\n",
    "        adata = adata[(~adata.obs.is_doublet)].copy()\n",
    "\n",
    "        print(f\"Number of cells after filtering of doublets: {adata.n_obs}\")\n",
    "        adata.write(main_path+\"datanew/\"+j+\".h5ad\")\n",
    "\n",
    "for a in raw_tiled_10x_mtx:\n",
    "    # For tiled TXT,CSV format\n",
    "    dataset_id = a\n",
    "    file_li = list(os.walk(main_path+dataset_id+\"_RAW/\"))[0][2]\n",
    "    ra = range(int(len(file_li)/3))\n",
    "    print(\"\\033[33mReading dataset {} ...\\033[0m\".format(a))\n",
    "    for o in ra:\n",
    "        prefix = file_li[o*3].split(\".\")[0][:-8]\n",
    "        adata = sc.read_10x_mtx(main_path+dataset_id+\"_RAW/\", prefix=prefix, gex_only=\"False\")\n",
    "        j = prefix.split(\"_\")[0]\n",
    "        print(\"\\033[33mReading sample {} ...\\033[0m\".format(j))\n",
    "        \n",
    "        # convert ensembl ids to gene symbols\n",
    "        new_var = list(adata.var_names)\n",
    "        if j in human_datasets:\n",
    "            new_var = [(\"MT-\"+i.split(\"MT\")[1]) if (i.startswith(\"MT\") and not i.startswith(\"MT-\")) else i for i in new_var] # turn \"MTXXX\" into \"MT-XXX\"\n",
    "            for i in range(len(new_var)):\n",
    "                try:\n",
    "                    if new_var[i][:4]==\"ENSG\" and pd.isna(hen2sy[new_var[i]])==False:\n",
    "                        new_var[i] = hen2sy[new_var[i]]\n",
    "                except:\n",
    "                    continue\n",
    "        else:\n",
    "            new_var = [(\"mt-\"+i.split(\"mt\")[1]) if (i.startswith(\"mt\") and not i.startswith(\"mt-\")) else i for i in new_var] # turn \"mtXXX\" into \"mt-XXX\"\n",
    "            for i in range(len(new_var)):\n",
    "                # if mouse mito gene symbols were in upper case, turn into lower case with \"mt-\"\n",
    "                if new_var[i] in [n.upper() for n in [\"Tf\",\"Rnr1\",\"Tv\",\"Rnr2\",\"Tl1\",\"Nd1\",\"Ti\",\"Tq\",\"Tm\",\"Nd2\",\"Tw\",\"Ta\",\"Tn\",\n",
    "                                                        \"Tc\",\"Ty\",\"Co1\",\"Ts1\",\"Td\",\"Co2\",\"Tk\",\"Atp8\",\"Atp6\",\"Co3\",\"Tg\",\"Nd3\",\"Tr\",\n",
    "                                                        \"Nd4l\",\"Nd4\",\"Th\",\"Ts2\",\"Tl2\",\"Nd5\",\"Nd6\",\"Te\",\"Cytb\",\"Tt\",\"Tp\"]]:\n",
    "                    new_var[i] = \"mt-\" + new_var[i][0] + new_var[i][1:].lower()\n",
    "                # turn ensembl id into gene symbol\n",
    "                try:\n",
    "                    if new_var[i][:7]==\"ENSMUSG\" and pd.isna(men2sy[new_var[i]])==False:\n",
    "                        new_var[i] = men2sy[new_var[i]]\n",
    "                except:\n",
    "                    continue\n",
    "        adata.var_names = pd.Index(new_var)\n",
    "        \n",
    "        # remove low sequencing quality cells\n",
    "        adata.var_names_make_unique()\n",
    "        adata.obs_names_make_unique()\n",
    "        sc.pp.filter_cells(adata, min_genes=200)\n",
    "        if j in human_datasets:\n",
    "            adata.var[\"mt\"] = adata.var_names.str.startswith(\"MT-\")\n",
    "        else:\n",
    "            adata.var[\"mt\"] = adata.var_names.str.startswith(\"mt-\") | adata.var_names.isin([n.upper() for n in [\"Tf\",\"Rnr1\",\"Tv\",\"Rnr2\",\"Tl1\",\"Nd1\",\"Ti\",\"Tq\",\"Tm\",\"Nd2\",\"Tw\",\"Ta\",\"Tn\",\n",
    "                                                                                                                \"Tc\",\"Ty\",\"Co1\",\"Ts1\",\"Td\",\"Co2\",\"Tk\",\"Atp8\",\"Atp6\",\"Co3\",\"Tg\",\"Nd3\",\"Tr\",\n",
    "                                                                                                                \"Nd4l\",\"Nd4\",\"Th\",\"Ts2\",\"Tl2\",\"Nd5\",\"Nd6\",\"Te\",\"Cytb\",\"Tt\",\"Tp\"]])\n",
    "        sc.pp.calculate_qc_metrics(\n",
    "            adata, qc_vars=[\"mt\"], inplace=True, percent_top=[20], log1p=True\n",
    "        )\n",
    "        adata.obs[\"outlier\"] = (\n",
    "            is_outlier(adata, \"log1p_total_counts\", 5)\n",
    "            | is_outlier(adata, \"log1p_n_genes_by_counts\", 5)\n",
    "            | is_outlier(adata, \"pct_counts_in_top_20_genes\", 5)\n",
    "        )\n",
    "        adata.obs.outlier.value_counts()\n",
    "        adata.obs[\"mt_outlier\"] = is_outlier(adata, \"pct_counts_mt\", 5)\n",
    "        adata.obs.mt_outlier.value_counts()\n",
    "        adata.obs[\"is_outlier\"] = [\"outlier\" if i else \"non-outlier\" for i in adata.obs[\"outlier\"]]\n",
    "        adata.obs[\"is_mt_outlier\"] = [\"mt_outlier\" if i else \"non-mt_outlier\" for i in adata.obs[\"mt_outlier\"]]\n",
    "        sc.pl.violin(adata, [\"pct_counts_mt\"], groupby = \"is_mt_outlier\")\n",
    "        \n",
    "        print(f\"Total number of cells: {adata.n_obs}\")\n",
    "        adata = adata[(~adata.obs.outlier)\\\n",
    "                    & (~adata.obs.mt_outlier)\\\n",
    "                    ].copy()\n",
    "        \n",
    "        print(f\"Number of cells after filtering of low quality cells: {adata.n_obs}\")\n",
    "        counts_matrix = adata.X\n",
    "        ro.globalenv[\"adata\"] = counts_matrix.T.todense()\n",
    "        ro.r('doublets <- scDblFinder(sce = as.matrix(adata), verbose=FALSE)')\n",
    "        doublet_scores = ro.r('doublets$scDblFinder.score').T\n",
    "        doublets = ro.r('doublets$scDblFinder.class')\n",
    "        adata.obs[\"doublet_score\"] = doublet_scores\n",
    "        adata.obs[\"doublet\"] = doublets\n",
    "        adata.obs[\"doublet\"] = [\"doublet\" if i!=1 else \"singlet\" for i in adata.obs[\"doublet\"]]\n",
    "        adata.obs[\"is_doublet\"] = [True if i==\"doublet\" else False for i in adata.obs[\"doublet\"]]\n",
    "        \n",
    "        print(f\"Total number of cells: {adata.n_obs}\")\n",
    "        adata = adata[(~adata.obs.is_doublet)].copy()\n",
    "\n",
    "        print(f\"Number of cells after filtering of doublets: {adata.n_obs}\")\n",
    "        adata.write(main_path+\"datanew/\"+j+\".h5ad\")\n",
    "\n",
    "for a in raw_h5:\n",
    "    # For h5 format\n",
    "    dataset_id = a\n",
    "    path_li = list(os.walk(main_path+dataset_id+\"_RAW/\"))[0][2]\n",
    "    print(\"\\033[33mReading dataset {} ...\\033[0m\".format(a))\n",
    "    for j in path_li:\n",
    "        print(\"\\033[33mReading sample {} ...\\033[0m\".format(j))\n",
    "        adata = sc.read_10x_h5(main_path+dataset_id+\"_RAW/\"+j)\n",
    "        j = j.split(\"_\")[0]\n",
    "        \n",
    "        # convert ensembl ids to gene symbols\n",
    "        new_var = list(adata.var_names)\n",
    "        if j in human_datasets:\n",
    "            new_var = [(\"MT-\"+i.split(\"MT\")[1]) if (i.startswith(\"MT\") and not i.startswith(\"MT-\")) else i for i in new_var] # turn \"MTXXX\" into \"MT-XXX\"\n",
    "            for i in range(len(new_var)):\n",
    "                try:\n",
    "                    if new_var[i][:4]==\"ENSG\" and pd.isna(hen2sy[new_var[i]])==False:\n",
    "                        new_var[i] = hen2sy[new_var[i]]\n",
    "                except:\n",
    "                    continue\n",
    "        else:\n",
    "            new_var = [(\"mt-\"+i.split(\"mt\")[1]) if (i.startswith(\"mt\") and not i.startswith(\"mt-\")) else i for i in new_var] # turn \"mtXXX\" into \"mt-XXX\"\n",
    "            for i in range(len(new_var)):\n",
    "                # if mouse mito gene symbols were in upper case, turn into lower case with \"mt-\"\n",
    "                if new_var[i] in [n.upper() for n in [\"Tf\",\"Rnr1\",\"Tv\",\"Rnr2\",\"Tl1\",\"Nd1\",\"Ti\",\"Tq\",\"Tm\",\"Nd2\",\"Tw\",\"Ta\",\"Tn\",\n",
    "                                                        \"Tc\",\"Ty\",\"Co1\",\"Ts1\",\"Td\",\"Co2\",\"Tk\",\"Atp8\",\"Atp6\",\"Co3\",\"Tg\",\"Nd3\",\"Tr\",\n",
    "                                                        \"Nd4l\",\"Nd4\",\"Th\",\"Ts2\",\"Tl2\",\"Nd5\",\"Nd6\",\"Te\",\"Cytb\",\"Tt\",\"Tp\"]]:\n",
    "                    new_var[i] = \"mt-\" + new_var[i][0] + new_var[i][1:].lower()\n",
    "                # turn ensembl id into gene symbol\n",
    "                try:\n",
    "                    if new_var[i][:7]==\"ENSMUSG\" and pd.isna(men2sy[new_var[i]])==False:\n",
    "                        new_var[i] = men2sy[new_var[i]]\n",
    "                except:\n",
    "                    continue\n",
    "        adata.var_names = pd.Index(new_var)\n",
    "        \n",
    "        # remove low sequencing quality cells\n",
    "        adata.var_names_make_unique()\n",
    "        adata.obs_names_make_unique()\n",
    "        sc.pp.filter_cells(adata, min_genes=200)\n",
    "        if j in human_datasets:\n",
    "            adata.var[\"mt\"] = adata.var_names.str.startswith(\"MT-\")\n",
    "        else:\n",
    "            adata.var[\"mt\"] = adata.var_names.str.startswith(\"mt-\") | adata.var_names.isin([n.upper() for n in [\"Tf\",\"Rnr1\",\"Tv\",\"Rnr2\",\"Tl1\",\"Nd1\",\"Ti\",\"Tq\",\"Tm\",\"Nd2\",\"Tw\",\"Ta\",\"Tn\",\n",
    "                                                                                                                \"Tc\",\"Ty\",\"Co1\",\"Ts1\",\"Td\",\"Co2\",\"Tk\",\"Atp8\",\"Atp6\",\"Co3\",\"Tg\",\"Nd3\",\"Tr\",\n",
    "                                                                                                                \"Nd4l\",\"Nd4\",\"Th\",\"Ts2\",\"Tl2\",\"Nd5\",\"Nd6\",\"Te\",\"Cytb\",\"Tt\",\"Tp\"]])\n",
    "        sc.pp.calculate_qc_metrics(\n",
    "            adata, qc_vars=[\"mt\"], inplace=True, percent_top=[20], log1p=True\n",
    "        )\n",
    "        adata.obs[\"outlier\"] = (\n",
    "            is_outlier(adata, \"log1p_total_counts\", 5)\n",
    "            | is_outlier(adata, \"log1p_n_genes_by_counts\", 5)\n",
    "            | is_outlier(adata, \"pct_counts_in_top_20_genes\", 5)\n",
    "        )\n",
    "        adata.obs.outlier.value_counts()\n",
    "        adata.obs[\"mt_outlier\"] = is_outlier(adata, \"pct_counts_mt\", 5)\n",
    "        adata.obs.mt_outlier.value_counts()\n",
    "        adata.obs[\"is_outlier\"] = [\"outlier\" if i else \"non-outlier\" for i in adata.obs[\"outlier\"]]\n",
    "        adata.obs[\"is_mt_outlier\"] = [\"mt_outlier\" if i else \"non-mt_outlier\" for i in adata.obs[\"mt_outlier\"]]\n",
    "        sc.pl.violin(adata, [\"pct_counts_mt\"], groupby = \"is_mt_outlier\")\n",
    "        \n",
    "        print(f\"Total number of cells: {adata.n_obs}\")\n",
    "        adata = adata[(~adata.obs.outlier)\\\n",
    "                    & (~adata.obs.mt_outlier)\\\n",
    "                    ].copy()\n",
    "        \n",
    "        print(f\"Number of cells after filtering of low quality cells: {adata.n_obs}\")\n",
    "        counts_matrix = adata.X\n",
    "        try:\n",
    "            ro.globalenv[\"adata\"] = counts_matrix.T.todense()\n",
    "        except:\n",
    "            ro.globalenv[\"adata\"] = counts_matrix.T\n",
    "        ro.r('doublets <- scDblFinder(sce = as.matrix(adata), verbose=FALSE)')\n",
    "        doublet_scores = ro.r('doublets$scDblFinder.score').T\n",
    "        doublets = ro.r('doublets$scDblFinder.class')\n",
    "        adata.obs[\"doublet_score\"] = doublet_scores\n",
    "        adata.obs[\"doublet\"] = doublets\n",
    "        adata.obs[\"doublet\"] = [\"doublet\" if i!=1 else \"singlet\" for i in adata.obs[\"doublet\"]]\n",
    "        adata.obs[\"is_doublet\"] = [True if i==\"doublet\" else False for i in adata.obs[\"doublet\"]]\n",
    "        sc.pl.violin(adata, [\"log1p_total_counts\",\"log1p_n_genes_by_counts\",\"pct_counts_in_top_20_genes\"], groupby=\"doublet\")\n",
    "        print(f\"Total number of cells: {adata.n_obs}\")\n",
    "        adata = adata[(~adata.obs.is_doublet)].copy()\n",
    "\n",
    "        print(f\"Number of cells after filtering of doublets: {adata.n_obs}\")\n",
    "        adata.write(main_path+\"datanew/\"+j+\".h5ad\")\n",
    "\n",
    "for a in raw_h5ad:\n",
    "    # For h5ad format\n",
    "    dataset_id = a\n",
    "    path_li = list(os.walk(main_path+dataset_id+\"_RAW/\"))[0][2]\n",
    "    print(\"\\033[33mReading dataset {} ...\\033[0m\".format(a))\n",
    "    for j in path_li:\n",
    "        print(\"\\033[33mReading sample {} ...\\033[0m\".format(j))\n",
    "        adata = sc.read_h5ad(main_path+dataset_id+\"_RAW/\"+j)\n",
    "        j = j.split(\"_\")[0]\n",
    "        \n",
    "        # convert ensembl ids to gene symbols\n",
    "        new_var = list(adata.var_names)\n",
    "        if j in human_datasets:\n",
    "            new_var = [(\"MT-\"+i.split(\"MT\")[1]) if (i.startswith(\"MT\") and not i.startswith(\"MT-\")) else i for i in new_var] # turn \"MTXXX\" into \"MT-XXX\"\n",
    "            for i in range(len(new_var)):\n",
    "                try:\n",
    "                    if new_var[i][:4]==\"ENSG\" and pd.isna(hen2sy[new_var[i]])==False:\n",
    "                        new_var[i] = hen2sy[new_var[i]]\n",
    "                except:\n",
    "                    continue\n",
    "        else:\n",
    "            new_var = [(\"mt-\"+i.split(\"mt\")[1]) if (i.startswith(\"mt\") and not i.startswith(\"mt-\")) else i for i in new_var] # turn \"mtXXX\" into \"mt-XXX\"\n",
    "            for i in range(len(new_var)):\n",
    "                # if mouse mito gene symbols were in upper case, turn into lower case with \"mt-\"\n",
    "                if new_var[i] in [n.upper() for n in [\"Tf\",\"Rnr1\",\"Tv\",\"Rnr2\",\"Tl1\",\"Nd1\",\"Ti\",\"Tq\",\"Tm\",\"Nd2\",\"Tw\",\"Ta\",\"Tn\",\n",
    "                                                        \"Tc\",\"Ty\",\"Co1\",\"Ts1\",\"Td\",\"Co2\",\"Tk\",\"Atp8\",\"Atp6\",\"Co3\",\"Tg\",\"Nd3\",\"Tr\",\n",
    "                                                        \"Nd4l\",\"Nd4\",\"Th\",\"Ts2\",\"Tl2\",\"Nd5\",\"Nd6\",\"Te\",\"Cytb\",\"Tt\",\"Tp\"]]:\n",
    "                    new_var[i] = \"mt-\" + new_var[i][0] + new_var[i][1:].lower()\n",
    "                # turn ensembl id into gene symbol\n",
    "                try:\n",
    "                    if new_var[i][:7]==\"ENSMUSG\" and pd.isna(men2sy[new_var[i]])==False:\n",
    "                        new_var[i] = men2sy[new_var[i]]\n",
    "                except:\n",
    "                    continue\n",
    "        adata.var_names = pd.Index(new_var)\n",
    "        \n",
    "        # remove low sequencing quality cells\n",
    "        adata.var_names_make_unique()\n",
    "        adata.obs_names_make_unique()\n",
    "        sc.pp.filter_cells(adata, min_genes=200)\n",
    "        if j in human_datasets:\n",
    "            adata.var[\"mt\"] = adata.var_names.str.startswith(\"MT-\")\n",
    "        else:\n",
    "            adata.var[\"mt\"] = adata.var_names.str.startswith(\"mt-\") | adata.var_names.isin([n.upper() for n in [\"Tf\",\"Rnr1\",\"Tv\",\"Rnr2\",\"Tl1\",\"Nd1\",\"Ti\",\"Tq\",\"Tm\",\"Nd2\",\"Tw\",\"Ta\",\"Tn\",\n",
    "                                                                                                                \"Tc\",\"Ty\",\"Co1\",\"Ts1\",\"Td\",\"Co2\",\"Tk\",\"Atp8\",\"Atp6\",\"Co3\",\"Tg\",\"Nd3\",\"Tr\",\n",
    "                                                                                                                \"Nd4l\",\"Nd4\",\"Th\",\"Ts2\",\"Tl2\",\"Nd5\",\"Nd6\",\"Te\",\"Cytb\",\"Tt\",\"Tp\"]])\n",
    "        sc.pp.calculate_qc_metrics(\n",
    "            adata, qc_vars=[\"mt\"], inplace=True, percent_top=[20], log1p=True\n",
    "        )\n",
    "        adata.obs[\"outlier\"] = (\n",
    "            is_outlier(adata, \"log1p_total_counts\", 5)\n",
    "            | is_outlier(adata, \"log1p_n_genes_by_counts\", 5)\n",
    "            | is_outlier(adata, \"pct_counts_in_top_20_genes\", 5)\n",
    "        )\n",
    "        adata.obs.outlier.value_counts()\n",
    "        adata.obs[\"mt_outlier\"] = is_outlier(adata, \"pct_counts_mt\", 5)\n",
    "        adata.obs.mt_outlier.value_counts()\n",
    "        adata.obs[\"is_outlier\"] = [\"outlier\" if i else \"non-outlier\" for i in adata.obs[\"outlier\"]]\n",
    "        adata.obs[\"is_mt_outlier\"] = [\"mt_outlier\" if i else \"non-mt_outlier\" for i in adata.obs[\"mt_outlier\"]]\n",
    "        sc.pl.violin(adata, [\"pct_counts_mt\"], groupby = \"is_mt_outlier\")\n",
    "        \n",
    "        print(f\"Total number of cells: {adata.n_obs}\")\n",
    "        adata = adata[(~adata.obs.outlier)\\\n",
    "                    & (~adata.obs.mt_outlier)\\\n",
    "                    ].copy()\n",
    "        \n",
    "        print(f\"Number of cells after filtering of low quality cells: {adata.n_obs}\")\n",
    "        counts_matrix = adata.X\n",
    "        try:\n",
    "            ro.globalenv[\"adata\"] = counts_matrix.T.todense()\n",
    "        except:\n",
    "            ro.globalenv[\"adata\"] = counts_matrix.T\n",
    "        ro.r('doublets <- scDblFinder(sce = as.matrix(adata), verbose=FALSE)')\n",
    "        doublet_scores = ro.r('doublets$scDblFinder.score').T\n",
    "        doublets = ro.r('doublets$scDblFinder.class')\n",
    "        adata.obs[\"doublet_score\"] = doublet_scores\n",
    "        adata.obs[\"doublet\"] = doublets\n",
    "        adata.obs[\"doublet\"] = [\"doublet\" if i!=1 else \"singlet\" for i in adata.obs[\"doublet\"]]\n",
    "        adata.obs[\"is_doublet\"] = [True if i==\"doublet\" else False for i in adata.obs[\"doublet\"]]\n",
    "        \n",
    "        print(f\"Total number of cells: {adata.n_obs}\")\n",
    "        adata = adata[(~adata.obs.is_doublet)].copy()\n",
    "\n",
    "        print(f\"Number of cells after filtering of doublets: {adata.n_obs}\")\n",
    "        adata.write(main_path+\"datanew/\"+j+\".h5ad\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "sc",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
